File: 03-Strings/valid_anagram.cpp
/*
Problem Name: Valid Anagram
Platform: LeetCode
Difficulty: Easy

Approach 1: Sorting both strings and comparing (O(n log n) time).
Approach 2: Character frequency count using hash map or fixed array of size 26.
Optimal Approach: Single hash map / array frequency counter.

Time Complexity: O(n)
Space Complexity: O(1) - since alphabet size is fixed to 26

Pattern: Hashing / Frequency Map
*/

#include <iostream>
#include <string>
#include <unordered_map>
#include <cassert>

class Solution {
public:
    bool isAnagram(std::string s, std::string t) {
        if (s.length() != t.length()) {
            return false;
        }
        
        std::unordered_map<char, int> char_counts;
        for (char c : s) {
            char_counts[c]++;
        }
        
        for (char c : t) {
            if (char_counts.find(c) == char_counts.end()) {
                return false;
            }
            char_counts[c]--;
            if (char_counts[c] == 0) {
                char_counts.erase(c);
            }
        }
        
        return char_counts.empty();
    }
};

int main() {
    Solution sol;
    assert(sol.isAnagram("anagram", "nagaram") == true);
    assert(sol.isAnagram("rat", "car") == false);
    std::cout << "Valid Anagram C++: All tests passed! 🎉" << std::endl;
    return 0;
}

File: 03-Strings/valid_anagram.py
# """
# Problem Name: Valid Anagram
# Platform: LeetCode
# Difficulty: Easy
#
# Approach 1: Sort strings s and t and check equality. O(n log n) time.
# Approach 2: Hash map tracking characters frequency of s and subtracting for t.
# Optimal Approach: Hash map frequency table check.
#
# Time Complexity: O(n)
# Space Complexity: O(1) - since character space is fixed at 26
#
# Pattern: Hashing / Frequency Map
# """

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        char_counts = {}
        for char in s:
            char_counts[char] = char_counts.get(char, 0) + 1
            
        for char in t:
            if char not in char_counts:
                return False
            char_counts[char] -= 1
            if char_counts[char] == 0:
                del char_counts[char]
                
        return len(char_counts) == 0

if __name__ == "__main__":
    sol = Solution()
    assert sol.isAnagram("anagram", "nagaram") is True
    assert sol.isAnagram("rat", "car") is False
    print("Valid Anagram Python: All tests passed! 🎉")
